### PR TITLE
feat: Add lotus-bench cli option to stress test any binary

### PR DESCRIPTION
**Pull request by fridrik01**
## Related Issues
Written to help with solving https://github.com/filecoin-project/lotus/issues/11251
Extension to the RPC benchmark introduced in https://github.com/filecoin-project/lotus/pull/10761

## Proposed Changes
This PR extends the RPC benchmark to support running any binary and analyze their latency and histogram distribution, track most common errors, perform stress testing under different concurrency levels and see how it works under different QPS. 

## Test plan

Lets say you want to understand how `lotus-shed mpool miner-select-messages` behaves over time. 

You can use this new `lotus-bench cli` tool to achieve this, for example the following command runs the command using a single thread (concurrency=1) at 1 qps and provides a realtime updated latency/histogram UI and it runs for 24 hours. 

`lotus-bench cli  --cmd='lotus-shed mpool miner-select-messages' --concurrency=1 --qps=1 --watch=5s --duration=24h`

It will produce something like:
image

## Additional Info
This works exactly the same as the rpc benchmark, which has a detailed summary (see [PR](https://github.com/filecoin-project/lotus/pull/10761))

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: cmd/lotus-bench/cli.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+* Gives a nice reporting summary of the stress testing of each command (including latency distribution, histogram and more)
+* Easy to use
+
+To use this benchmark you must specify the commands you want to test using the --cmd options, the format of it is:
+
+  --cmd=CMD[:CONCURRENCY][:QPS] where only NAME is required.
+
+Here are some real examples:
+  lotus-bench cli --cmd='lotus-shed mpool miner-select-messages' // runs the command with default concurrency and qps
+  lotus-bench cli --cmd='lotus-shed mpool miner-select-messages:3'  // override concurrency to 3
```
Comment (Stebalien) [inline cmd/lotus-bench/cli.go:34]: Using `:` could cause issues down the road (for any command including a literal `:`. Why not just rely on the flags?

Also, for the command, I'd suggest using positional parameters. I.e., "the rest of the args".

**Review by Stebalien — COMMENTED** [review]
This seems generally useful but an odd addition to `lotus-bench` given that it's general-purpose.

It looks like you'd be better off with something like https://github.com/sharkdp/hyperfine (available in pretty much every package manager)

**Comment by fridrik01** [comment]
> This seems generally useful but an odd addition to `lotus-bench` given that it's general-purpose.
> 
> It looks like you'd be better off with something like https://github.com/sharkdp/hyperfine (available in pretty much every package manager)

Yes, this could be an independent general purpose tool but I put it here for a lack of a better place to put it as it already shares many of the same code when stress testing our RPC endpoints, and this would enable us to do the same by stress testing lotus backend through our different cli commands.

I am not familiar with hyperfine but this does seem different enough as its based on generating specific workloads for a  certain duration where you can tune the concurrency/qps for each independently and then get simple histogram/latency distribution.

File: cmd/lotus-bench/cli.go (hunk 1 above)
Comment (arajasek) [reply cmd/lotus-bench/cli.go:34]: +1 here.

**Review by arajasek — APPROVED** [review]
This LGTM, and I'm personally happy to land as-is. @magik6k do you wanna weigh in too?

**Comment by magik6k** [comment]
> @magik6k do you wanna weigh in too?

Didn't review, but I'm fine with this landing

**Pull request merged**